client/src/java/com/zimbra/client/ZMessage.java

```diff
@@ -561,13 +566,33 @@ public MailItemType getMailItemType() {
         return MailItemType.MESSAGE;
     }
 
+    /**
+     * @return the UID the item is referenced by in the IMAP server.  Returns <tt>0</tt> for items that require
+     * renumbering because of moves.
+     * The "IMAP UID" will be the same as the item ID unless the item has been moved after the mailbox owner's first
+     * IMAP session. */
     @Override
     public int getImapUid() {
-        throw new UnsupportedOperationException("ZMessage method not supported yet");
+        if ((null != imapUid) && (imapUid >= 0)) {
+            return imapUid;
+        }
+        ZimbraLog.mailbox.debug("ZMessage getImapUid() - regetting UID");
+        ZMessage zm = null;
+        try {
+            zm = getMailbox().getMessageById(mId);
```

Comment:
The cache lookup misses if different things were requested - I need another look to check that the lookup fails if Uid wasn't asked for before.